### PR TITLE
Fix problems in scene editor and hierarchy g UI

### PR DESCRIPTION
**Pull request by Erdios**
~~❗❗Merge [Inspector Window to actually add/remove/update components of a game object branch](https://github.com/kablouser/PlatinumEngine/pull/24) first before merging this branch.~~ merged

EDIT from Matt:
Another way to reference a PR is by simply putting #24 :)

### CODE REVIEW

File: Include/SceneManager/HierarchyWindow.h [hunk 1]
```diff
@@ -36,15 +36,23 @@ namespace PlatinumEngine
 		// ---> PARAMETER
 		GameObject* selectedGameObject;
 
+
```
Comment (Matt-Cumber) [inline Include/SceneManager/HierarchyWindow.h:39]:
```suggestion
```

File: Include/SceneManager/HierarchyWindow.h [hunk 2]
```diff
@@ -36,15 +36,23 @@ namespace PlatinumEngine
...
 	private:
+		// ---> ENUM
+		enum ModeForDraggingBehavior{_noneMode, _hierarchyMode, _orderMode};
+
```
Comment (Matt-Cumber) [inline Include/SceneManager/HierarchyWindow.h:43]:
```suggestion
```

File: Include/SceneManager/HierarchyWindow.h [hunk 3]
```diff
@@ -36,15 +36,23 @@ namespace PlatinumEngine
...
 		// ---> FUNCTION
 		/**
 		 * Function to display nested tree node for all the game objects in the scene
 		 * @param gameObject : the game object that is processing
 		 * @param scene : scene object reference that carries the game object hierarchy structure
 		 */
-		void DisplayTreeNote(GameObject* gameObject, Scene& scene);
+		void DisplayTreeNote(GameObject* gameObject, Scene& scene, ModeForDraggingBehavior modeForDraggingBehavior);
+
+
```
Comment (Matt-Cumber) [inline Include/SceneManager/HierarchyWindow.h:53]:
```suggestion
```

File: Include/SceneManager/Scene.h [hunk 4]
```diff
@@ -56,14 +59,28 @@ namespace PlatinumEngine
 
 		GameObject* GetRootGameObject(size_t index);
 
+		// Returns boolean to check if moving successes or not
+		// targetObject : a target object of which the moved object wants to move in front of
+		// movedGameObject : an object that want to be moved
+		bool MoveRootGameObjectPositionInList(GameObject* targetObject, GameObject* movedGameObject);
+
+
```
Comment (Matt-Cumber) [inline Include/SceneManager/Scene.h:67]:
```suggestion
```

File: Include/SceneManager/Scene.h [hunk 5]
```diff
@@ -103,6 +120,13 @@ namespace PlatinumEngine
 		 */
 		void Render(Renderer& renderer);
 
+
+
+
+
+
+
+
```
Comment (Matt-Cumber) [inline Include/SceneManager/Scene.h:129]:
```suggestion
```

File: Source/ComponentComposition/RenderComponent.cpp [hunk 6]
```diff
@@ -7,7 +7,7 @@ namespace PlatinumEngine
 {
 	RenderComponent::RenderComponent()
 	{
-		//LoadMesh("C:/Users/asus/Desktop/course/games101/Assignment3/models/rock/rock.obj");
+		//LoadMesh("D:/Download/fourCubes.obj");
```
Comment (Matt-Cumber) [inline Source/ComponentComposition/RenderComponent.cpp:10]:
```suggestion
```

File: Source/SceneManager/HierarchyWindow.cpp [hunk 7]
```diff
@@ -69,13 +104,58 @@ namespace PlatinumEngine
...
+
+			ImGui::InvisibleButton(" ", ImVec2(windowSize.x,10));
+
+			// Add Drag and Drop Events
+			if (ImGui::BeginDragDropSource(ImGuiDragDropFlags_None))
+			{
+
+				// Set payload to carry the game object
+				ImGui::SetDragDropPayload("Demo", nullptr, sizeof(void*));
+				std::cout<<"test"<<std::endl;
```
Comment (Matt-Cumber) [inline Source/SceneManager/HierarchyWindow.cpp:138]:
```suggestion
```

File: Source/main.cpp [hunk 8]
```diff
@@ -84,16 +84,22 @@ int main(int, char**)
...
-		PlatinumEngine::InspectorWindow inspectorWindow;
+		PlatinumEngine::InspectorWindow inspectorWindow(scene);
 
 		PlatinumEngine::WindowManager windowManager;
 
 
+
 		//---------TEST-----------
 
+		//Mesh mesh(vertices, indices);
```
Comment (Matt-Cumber) [inline Source/main.cpp:95]:
```suggestion
```

File: Source/SceneManager/Scene.cpp [hunk 9]
```diff
@@ -92,7 +92,33 @@ namespace PlatinumEngine
...
+		auto movedGameObjectIterator = std::find(_rootGameObjects.begin(), _rootGameObjects.end(), movedGameObject);
+
+		// use rotate to move the item in front of the target object
+		if(targetGameObjectIterator!= _rootGameObjects.end() && movedGameObjectIterator != _rootGameObjects.end())
+		{
+			if(movedGameObjectIterator < targetGameObjectIterator)
+			std::rotate(movedGameObjectIterator, movedGameObjectIterator+1, targetGameObjectIterator+1);
+
+			else if (movedGameObjectIterator > targetGameObjectIterator)
+			std::rotate(targetGameObjectIterator, movedGameObjectIterator, movedGameObjectIterator+1);
```
Comment (Matt-Cumber) [inline Source/SceneManager/Scene.cpp:110]: Would you be able to indent these

**Review by Matt-Cumber — CHANGES_REQUESTED** [review]
I've tried to ignore changes from the other PR and the loader changes (I'm not sure why they are here these should be in main already , but will hopefully be resolved by merging main into this branch). 

Code works well, and nice comments make it as easy to follow as it can be. Just need to tidy up the code a little :)

File: Source/SceneEditor/SceneEditor.cpp [hunk 10]
```diff
@@ -264,14 +243,14 @@ namespace PlatinumEngine{
...
-			RenderComponent renderComponent;
+
 			_renderer->Begin();
-			renderComponent.OnRender(*_scene, *_renderer);
 			_renderer->SetModelMatrix();
 			_renderer->SetViewMatrix(_camera.viewMatrix4);
 			_renderer->SetProjectionMatrix(_camera.projectionMatrix4);
 			_renderer->SetLightProperties();
+			// Update rendering information to renderer
+			_scene->Render(*_renderer);
```
Comment (ShawnWicky) [inline Source/SceneEditor/SceneEditor.cpp:253]:
```suggestion
                        // Update rendering information to renderer
                        _scene->Render(*_renderer);
			_renderer->SetModelMatrix();
			_renderer->SetViewMatrix(_camera.viewMatrix4);
			_renderer->SetProjectionMatrix(_camera.projectionMatrix4);
			_renderer->SetLightProperties();
```

**Pull request merged**